packages/babel-parser/src/parser/expression.js

```diff
@@ -1208,12 +1259,102 @@ export default class ExpressionParser extends LValParser {
           }
         }
       }
+
       // fall through
       default:
         throw this.unexpected();
     }
   }
 
+  // https://github.com/js-choi/proposal-hack-pipes
+  maybeParseTopicReference(): ?N.Expression {
+    const pipeProposal = this.getPluginOption("pipelineOperator", "proposal");
+
+    // `pipeProposal` is falsy when an input program
+    // contains a topic reference on its own,
+    // outside of a pipe expression,
+    // and without having turned on the pipelineOperator plugin.
+    if (pipeProposal) {
+      // A pipe-operator proposal is active.
+
+      const tokenType = this.state.type;
+
+      if (this.testTopicReferenceConfiguration(pipeProposal, tokenType)) {
+        // The token matches the plugin’s configuration.
+        // The token is therefore a topic reference.
+
+        const node = this.startNode();
+
+        // Determine the node type for the topic reference
+        // that is appropriate for the active pipe-operator proposal.
+        let nodeType;
+        if (pipeProposal === "smart") {
+          nodeType = "PipelinePrimaryTopicReference";
+        } else {
+          // The proposal must otherwise be "hack",
+          // as enforced by testTopicReferenceConfiguration.
+          nodeType = "TopicReference";
+        }
+
+        // Consume the token.
+        this.next();
+
+        // Register the topic reference so that its pipe body knows
+        // that its topic was used at least once.
+        this.registerTopicReference();
+
+        if (!this.topicReferenceIsAllowedInCurrentContext()) {
+          // The topic reference is not allowed in the current context:
+          // it is outside of a pipe body.
+          // Raise recoverable errors.
+          if (pipeProposal === "smart") {
+            this.raise(this.state.start, Errors.PrimaryTopicNotAllowed);
+          } else {
+            // In this case, `pipeProposal === "hack"` is true.
+            this.raise(this.state.start, Errors.PipeTopicUnbound);
+          }
+        }
+
+        return this.finishNode(node, nodeType);
+      } else {
+        // The token does not match the plugin’s configuration.
+        throw this.raise(
+          this.state.start,
+          Errors.PipeTopicUnconfiguredToken,
+          tokenType.label,
+        );
```

Comment:
I don’t think this will be a problem. The tokenizer consumes the `#[` preferentially to `#`, which means that `maybeParseTopicReference` will never be called.

I added a [test for combining `topicToken: "%"` with hash tuples](https://github.com/js-choi/babel/tree/proposal-hack-pipes-percent/packages/babel-parser/test/fixtures/experimental/pipeline-operator/hack-percent-proposal-hash-tuple) in the [`js-choi:proposal-hack-pipes-percent` branch](https://github.com/js-choi/babel/tree/proposal-hack-pipes-percent/). This test checks that `a |> #[0, %]` correctly parses.

Incidentally, we are [unlikely to choose `?` as a topic token](https://github.com/js-choi/proposal-hack-pipes/issues/2#issuecomment-849179767), so it may not be worth adding support for it to Babel. `%` and `#` are the current frontrunners.